ergo/distributions/point_density.py

```diff
@@ -0,0 +1,260 @@
+from dataclasses import dataclass
+
+from jax import nn
+import jax.numpy as np
+import numpy as onp
+from scipy.interpolate import interp1d
+
+from ergo.scale import Scale
+
+from . import constants
+from .distribution import Distribution
+from .optimizable import Optimizable
+
+
+@dataclass
+class PointDensity(Distribution, Optimizable):
+    """
+    A distribution specified through a number of density points.
+    """
+
+    normed_xs: np.DeviceArray
+    normed_densities: np.DeviceArray
+    bin_probs: np.DeviceArray
+    scale: Scale
+
+    def __init__(
+        self,
+        xs,
+        densities,
+        scale: Scale,
+        normalized=False,
+        traceable=False,
+        cumulative_normed_ps=None,
+    ):
+        if scale is None:
+            raise ValueError
+
+        self.scale = scale
+        init_np = np if traceable else onp
+
+        if normalized:
+            self.normed_xs = xs
+            self.normed_densities = densities
+
+        else:
+            self.normed_xs = scale.normalize_points(xs)
+            self.normed_densities = scale.normalize_densities(self.normed_xs, densities)
+
+        self.bin_probs = self.normed_densities * constants.bin_sizes
+
+        if cumulative_normed_ps is not None:
+            self.cumulative_normed_ps = cumulative_normed_ps
+
+        else:
+            self.cumulative_normed_ps = init_np.append(
+                init_np.array([0]), init_np.cumsum(self.bin_probs)
+            )
+        self.normed_log_densities = init_np.log(self.normed_densities)
```

Comment:
`normed_log_densities` doesn't seem to be used anywhere

Reply:
It is used in the smoothness condition, but it could perhaps be calculated there for efficiency reasons